README
・ このライブラリは libpasori (http://libpasori.sourceforge.jp/) を再構
   成し、若干の機能追加などを行なったものです。

・ NFC Reader/Writer Linux Driver ( http://core.dumped.cc/devel/nfc/index.ja.html )
   および
   libnfc ( http://www.libnfc.org/ )
   のソースコードも参考にしています。

・ Sony の PaSoRi RC-S320 および RC-S330 に対応しています。

・ 動作確認などは不十分な状態なので、各自の責任でご利用ください。


●コンパイルとインストール

./configure
make
make install


●udev の設定

Debian GNU/Linux lenny での設定例です。ディストリビューションやバージョ
ンにより設定が異なる可能性がありますので適宜読み替えて設定を行なってく
ださい。

1. 下記の内容を /etc/udev/pasori.rules として保存する。

ACTION!="add", GOTO="pasori_rules_end"
SUBSYSTEM=="usb_device", GOTO="pasori_rules_start"
SUBSYSTEM!="usb", GOTO="pasori_rules_end"
LABEL="pasori_rules_start"

ATTRS{idVendor}=="054c", ATTRS{idProduct}=="01bb", MODE="0664", GROUP="plugdev"
ATTRS{idVendor}=="054c", ATTRS{idProduct}=="02e1", MODE="0664", GROUP="plugdev"

LABEL="pasori_rules_end"


2. /etc/udev/rules.d/にシンボリック・リンクを張る。

cd /etc/udev/rules.d
ln -s ../pasori.rules 010_pasori.rules

ファイル名は適宜変更してください。

3. udevadm control --reload-rules を実行。


以上で plugdev グループに属しているユーザは pasori を利用できるようにな
るはずです。



●関数リファレンス

pasori *pasori_open(void);

概要
  デバイスファイルをオープンする。

返り値
  成功した場合 psori 型のポインタを返す。失敗すると NULL が返される。



int pasori_init(pasori *p)

概要
  PaSoRi の初期化を行なう。

引数
  p  pasori_open で取得した psori 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



void pasori_close(pasori *p);

概要
  pasori_reset() を実行した後、デバイスファイルをクローズする。

引数
  p  pasori_open で取得した psori 型のポインタ。



int pasori_send(pasori *p, uint8 *data, int *size);

概要
  PaSoRi にデータを送出する。

引数
  p    pasori_open で取得した psori 型のポインタ。
  data 送出するデータへのポインタ。
  size 送出するデータのサイズ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。


int pasori_recv(pasori *p, uint8 *data, int *size);

概要
  PaSoRi からデータを受け取る。

引数
  p    pasori_open で取得した psori 型のポインタ。
  data データを格納するバッファへのポインタ。
  size バッファのサイズへのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。
  size が参照している場所に格納したデータの長さが保存される。



int pasori_packet_write(pasori *p, uint8 *data, int *size)

概要
  PaSoRi にコマンドを送出する。内部で pasori_send を呼び出している。

引数
  p    pasori_open で取得した psori 型のポインタ。
  data 送出するデータへのポインタ。
  size 送出するデータのサイズへのポインタ。

返り値
  成功した場合 0 を返す。size が参照している場所に送出したデータの長さ
  が保存される。失敗すると 0 以外の数値が返される。



int pasori_packet_read(pasori * p, uint8 * data, int *size);

概要
  PaSoRi に送出したコマンドへの応答を取得する。内部で pasori_recv を呼
  び出している。

引数
  p    pasori_open で取得した psori 型のポインタ。
  data データを格納するバッファへのポインタ。
  size バッファのサイズへのポインタ。

返り値
  成功した場合 0 を返す。size が参照している場所に取得したデータの長さ
  が保存される。失敗すると 0 以外の数値が返される。


int pasori_write(pasori *p, uint8 *data, int *size);

概要
  FeliCa にデータを送信する。内部で pasori_packet_write を呼び出してい
  る。
  RC-S320 では PaSoRi2 コマンド 0x5C に対応する。

引数
  p    pasori_open で取得した psori 型のポインタ。
  data データを格納するバッファへのポインタ。
  size バッファのサイズへのポインタ。

返り値
  成功した場合 0 を返す。size が参照している場所に送信したデータの長さ
  が保存される。失敗すると 0 以外の数値が返される。


int pasori_read(pasori *p, uint8 *data, int *size);

概要
  FeliCa からの応答を受信する。内部で pasori_packet_read を呼び出してい
  る。

引数
  p    pasori_open で取得した psori 型のポインタ。
  data データを格納するバッファへのポインタ。
  size バッファのサイズへのポインタ。

返り値
  成功した場合 0 を返す。size が参照している場所に格納したデータの長さ
  が保存される。失敗すると 0 以外の数値が返される。


int pasori_reset(pasori * p);

概要
  RC-S320 では Pasori にリセットコマンドを送出する。(PaSoRi2 コマンド 0x54)
  RC-S330 では RF Off などの処理を行う。

引数
  p    pasori_open で取得した psori 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int pasori_version(pasori *p, int *v1, int *v2);

概要
  Pasori のファームウェアバージョンを取得する。

引数
  p  pasori_open で取得した psori 型のポインタ。
  v1 バージョンの上位番号を格納するためのポインタ。
  v2 バージョンの下位番号を格納するためのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。

バグ
  RC-S330 で正しい値が得られているかどうかは未確認。



int pasori_type(pasori *p);

概要
  Pasori のタイプを取得する。

引数
  p  pasori_open で取得した psori 型のポインタ。

返り値
  PASORI_TYPE_S310
  PASORI_TYPE_S320
  PASORI_TYPE_S330
  
  失敗すると -1 が返される。



int pasori_test(pasori *p, int code, uint8 *data, int *size, uint8 *rdata, int *rsize);

概要
  RC-S320 のみ有効。
  Pasori の自己診断を行なう。(PaSoRi2 コマンド 0x52)

引数
  p     pasori_open で取得した psori 型のポインタ。
  code  テストコード
  data  テストデータを格納するバッファへのポインタ。
  size  テストデータバッファのサイズへのポインタ。
  rdata 応答データを格納するバッファへのポインタ。
  rsize 応答データバッファのサイズへのポインタ。


返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int pasori_test_echo(pasori *p, uint8 *data, int *size);

概要
  RC-S320 のみ有効。
  Pasori のエコーバックテストを行なう。(テストコード 0x00)

引数
  p    pasori_open で取得した psori 型のポインタ。
  data テストデータを格納するバッファへのポインタ。
  size テストデータバッファのサイズへのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int pasori_test_eprom(pasori *p);

概要
  RC-S320 のみ有効。
  Pasori の EPROM テストを行なう。(テストコード 0x01)

引数
  p    pasori_open で取得した psori 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int pasori_test_ram(pasori *p);

概要
  RC-S320 のみ有効。
  Pasori の RAM テストを行なう。(テストコード 0x02)

引数
  p    pasori_open で取得した psori 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int pasori_test_cpu(pasori *p);

概要
  RC-S320 のみ有効。
  Pasori の CPU テストを行なう。(テストコード 0x03)

引数
  p    pasori_open で取得した psori 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int pasori_test_polling(pasori *p);

概要
  RC-S320 のみ有効。
  Pasori のポーリングテストを行なう。(テストコード 0x10)

引数
  p pasori_open で取得した psori 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



void pasori_set_timeout(pasori *p, int timeout);

概要
  タイムアウトを設定する。

引数
  p pasori_open で取得した psori 型のポインタ。
  timeout timeout の値 (usb_control_msg() で使用される)。



felica *felica_polling(pasori *p, uint16 systemcode, uint8 RFU, uint8 timeslot);

概要
  REQC (リクエストコマンド C 型) を送信し、以降の通信で必要なデータを取
  得する。

引数
  p pasori_open で取得した psori 型のポインタ。
  systemcode システムコード
              FELICA_POLLING_ANY   全てのシステムが反応する
              FELICA_POLLING_SUICA Suica
              FELICA_POLLING_EDY   Edy
  RFU 予備 (通常は 0 を指定)
  timeslot タイムスロットの最大値を指定する。詳細は JIS X 6319-4 を参照のこと。

返り値
  成功した場合 felica 型のポインタを返す。返された felica 型のポインタ
  は不要になったときに free(3) で開放する必要がある。 失敗すると NULL
  が返される。



int felica_get_idm(felica *f, uint8 *idm);

概要
  IDm (製造識別子) を取得する。

引数
  f   felica_polling で取得した felica 型のポインタ。
  idm IDm を格納するためのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int felica_get_pmm(felica *f, uint8 *pmm);

概要
  PMm (製造パラメタ) を取得する。

引数
  f   felica_polling で取得した felica 型のポインタ。
  pmm PMm を格納するためのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int felica_read(felica * f, int *n, felica_block_info *info, uint8 *data);

概要
  Read コマンドを送信する。

引数
  f    felica_polling で取得した felica 型のポインタ。
  n    ブロック数へのポインタ。
  info サービスコードリスト。
  data 応答データを格納するバッファへのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。

バグ
  RC-S330 では動作未確認。


int felica_read_single(felica * f, int servicecode, int mode, uint8 block, uint8 *data);

概要
  Read コマンドを送信する。ブロック数は 1 に固定。

引数
  f     felica_polling で取得した felica 型のポインタ。
  mode  アクセスモード
        0  直接アクセス
           循環順編成アクセス
           減算アクセス
        1  戻入れアクセス
  block ブロック番号
  data  応答データを格納するバッファへのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int felica_write(felica *f, int servicecode, int mode, uint8 addr, uint8 *data);

未実装



int felica_request_service(felica *f, int *n, uint16 *list, uint16 *data);

概要
  Request Service コマンドを送信する。

引数
  f    felica_polling で取得した felica 型のポインタ。
  n    ブロック数へのポインタ。
  list サービスコードリストまたはエリアコードリスト。
  data 応答データを格納するバッファへのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。

バグ
  RC-S330 では動作未確認。


int felica_request_response(felica *f, uint8 *mode);

概要
  Request Response コマンドを送信する。現在のモードを調べる。

引数
  f    felica_polling で取得した felica 型のポインタ。
  mode モードを格納するためのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。

バグ
  RC-S330 では動作未確認。


int felica_search_service(felica *f);

概要
  0xffffがサービスコードとして返却されるまで Search Service Code コマン
  ドを送信する。
  取得されたサービスコードまたはエリアコードの数は

  f->area_num, f->service_num

  に設定されます。  それぞれのサービスコードまたはエリアコードは

  f->area[n].code, f->area[n].attr, f->area[n].bin
  f->service[n].code, f->service[n].attr, f->service[n].bin

  で参照できます。code はサービス番号またはサービスファイルの最小ファイ
  ル識別子、attr はアクセス制御処理区分、 bin はファイル識別子全体を表
  します。

引数
  f    felica_polling で取得した felica 型のポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



int felica_request_system(felica *f, int *n, uint16 *data);

概要
  Request System Code コマンドを送信する。FeliCa のもつシステムコードの
  リストを得る。

引数
  f    felica_polling で取得した felica 型のポインタ。
  data 応答データを格納するバッファへのポインタ。

返り値
  成功した場合 0 を返す。失敗すると 0 以外の数値が返される。



定数

FELICA_IDM_LENGTH      IDm のサイズ
FELICA_PMM_LENGTH      PMm のサイズ
FELICA_BLOCK_LENGTH    ブロックのサイズ



戻り値

PASORI_ERR_PARM    引数に誤りがある。
PASORI_ERR_MEM     メモリの確保に失敗した。
PASORI_ERR_COM     通信エラー。
PASORI_ERR_DATA    データに異常がある。
PASORI_ERR_CHKSUM  チェックサムに失敗した。
PASORI_ERR_FORMAT  コマンドの形式が異常。
PASORI_ERR_TYPE    未対応のハードウェア。


●連絡先

ZXB01226@nifty.com （オリジナルソースコード）
rfujita@redhat.com （RPM化及びバージョン0.0.7をベースにしたgit化）

ChangeLog
2011-08-14  Hiroyuki Ito  <ZXB01226 at nifty.com>

	* src/pasori_command.c (INTERFACE_NUMBER): added.
	(pasori_close, open_usb): use "INTERFACE_NUMBER".
	(struct tag_pasori): change name of some members.
	(get_end_points): check endpoint direction for interrupt transfer.

2011-08-13  Hiroyuki Ito  <ZXB01226 at nifty.com>

	* src/pasori_command.c (tag_pasori): remove member "dev", add
	member "ep_interrupt".
	(get_end_points): use "libusb_get_device()" to get diveice.
	(get_end_points): check transfer type.
	(open_usb): use "libusb_open()" instead of
	"libusb_open_device_with_vid_pid()".
	(pasori_open): call "memset()" to initialize "pasori" struct.
	(pasori_send, pasori_recv): use "ep_interrupt" for interrupt
	transfer.

2011-08-12  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* tests/pasori_test.c (show_version, main): use "pasori_type()".

	* src/pasori_command.c: add definition of "tag_pasori".
	(pasori_type): addes.

	* src/libpafe.h: "tag_pasori" is incomplete type.

	* src/felica_command.c (felica_pasori_read, felica_polling): use
	"pasori_type()".

	* README: add document for "pasori_type()".

2011-08-11  Hiroyuki Ito  <ZXB01226 at nifty.com>

	* debian/control: fix dependance.

	* src/pasori_command.c: support libusb version 1.0.

	* src/libpafe.h: support libusb version 1.0.

	* configure.in: use pkg-config to check existance of libusb.

	* tests/felica_dump.c (dump_service): remove unused variables
	"area_count" and "servicecode_count".

	* src/pasori_command.c (pasori_version): remove unused variable
	"ofst".

2011-08-11  Ryo Fujita  <rfujita at redhat.com>

	* spec/libpafe.spec: added.

2009-10-09  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* tests/pasori_test.c (show_version, main): add code for
	RC-S310 (http://felicabalviewer.sourceforge.jp/linux.html).

	* src/libpafe.h (enum PASORI_TYPE):add
	"PASORI_TYPE_S310" (http://felicabalviewer.sourceforge.jp/linux.html).

	* src/felica_command.c (felica_pasori_read, felica_polling): add
	code for
	RC-S310 (http://felicabalviewer.sourceforge.jp/linux.html).

	* src/pasori_command.c (PASORIUSB_PRODUCT_S310, S310_INIT):
	add (http://felicabalviewer.sourceforge.jp/linux.html).
	(pasori_test, pasori_write, pasori_read, pasori_init)
	(pasori_reset, pasori_version, pasori_open, pasori_send)
	(pasori_recv): add code for
	RC-S310 (http://felicabalviewer.sourceforge.jp/linux.html).

	* debian/libpafe.udev: add definition for
	RC-S310 (http://felicabalviewer.sourceforge.jp/linux.html).

2009-07-20  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/pasori_command.c (bcd2int): add.
	(pasori_version): use "bcd2int()" when "p->type" is
	"PASORI_TYPE_S330".

2009-07-17  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* tests/pasori_test.c (show_version, main): check "p->type".

	* src/pasori_command.c (S330_RF_ANTENNA_ON, S330_RF_ANTENNA_OFF)
	(S330_GET_VERSION, S330_DESELECT, pasori_list_passive_target)
	(get_end_points): add.
	(pasori_init_test): return status.
	(pasori_write, pasori_read, pasori_init, pasori_reset)
	(pasori_version, pasori_open, pasori_send, pasori_recv)
	(pasori_test): check "p->type".
	(pasori_close): call "pasori_reset()" before close the usb device.

	* src/libpafe.h (enum PASORI_TYPE): add.
	(struct tag_pasori): add member "type".
	(PASORI_ERR_TYPE): add.

	* src/felica_command.c (_felica_pasori_read, felica_pasori_read):
	add.
	(felica_polling): use "pasori_list_passive_target()" when
	"pp->type" is "PASORI_TYPE_S330".
	(felica_search_service, felica_request_service)
	(felica_request_response, felica_request_system): use
	"felica_pasori_read()" instead of "pasri_read()".

2009-05-24  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/Makefile.am (libpafe_la_LDFLAGS): "-lusb" を追加。

2008-03-20  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/libpafe.h (FELICA_SERVICE_IRUCA_HISTORY): 追加

2008-02-14  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/pasori_command.c (checksum): Fix checksum calculation.

2008-01-29  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/libpafe.h (FELICA_SERVICE_EDY_HISTORY)
	(FELICA_SERVICE_EDY_NUMBER): 追加

2008-01-22  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* COPYING: BSD ライセンスに変更

2008-01-17  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/libpafe.h (FELICA_SERVICE_SUICA_IN_OUT)
	(FELICA_SERVICE_SUICA_HISTORY)
	(FELICA_CMD_AUTHENTICATION1, FELICA_ANS_AUTHENTICATION1)
	(FELICA_CMD_AUTHENTICATION2, FELICA_ANS_AUTHENTICATION2)
	(FELICA_CMD_READ_FROM_SECURE_FILE)
	(FELICA_ANS_READ_FROM_SECURE_FILE)
	(FELICA_CMD_WRITE_TO_SECURE_FILE)
	(FELICA_ANS_WRITE_TO_SECURE_FILE): 追加

2008-01-15  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* src/felica_command.c (felica_read_single): fix typo (felica_read_sigle から修正)
	* 一部の関数で返り値がおかしかったのを修正。

2008-01-14  Ito Hiroyuki  <ZXB01226 at nifty.com>

	* Version 0.0.1 公開


tests/felica_dump.c
/* $Id: felica_dump.c,v 1.6 2009-07-17 07:30:55 hito Exp $ */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "config.h"
#include "libpafe.h"

void
mydump(uint8 * p, int size)
{
  int i;
  for (i = 0; i != size; i++) {
    printf("%02X", p[i]);
  }
}

void
show_idminfo(uint8 * idm)
{
  time_t mtime;
  time_t ntime;
  struct tm fepoc;
  signed int d;

  fepoc.tm_sec = 0;
  fepoc.tm_min = 0;
  fepoc.tm_hour = 0;
  fepoc.tm_mday = 1;
  fepoc.tm_mon = 0;
  fepoc.tm_year = 100;
  fepoc.tm_isdst = 0;

  mtime = mktime(&fepoc);
  ntime = time(NULL);

  d = idm[4] + idm[5] * 0x100;
  if (d & 0x8000) {
    d = 0 - (0x10000 - d);
  }
  mtime += d * 24 * 60 * 60;
  fepoc = *localtime(&mtime);	/* FIXME:structure copy */

  printf("# lpdump : %s", ctime(&ntime));

  printf("# --- IDm info (FeliCa) ---\n");
  printf("# Manufacture Date = %d/%d/%d\n", fepoc.tm_year + 1900,
	 fepoc.tm_mon + 1, fepoc.tm_mday);
  printf("#               SN = %d\n", idm[6] + idm[7] * 0x100);
  printf("# Manufacture Code = %04X\n", idm[0] + idm[1] * 0x100);
  printf("#      Equip. Code = %04X\n", idm[2] + idm[3] * 0x100);
}

struct __tag__sinfo
{
  int c;
  char *d;
};

typedef struct __tag__sinfo sinfo;
const sinfo Sifo[] = {
  {0, " Area Code                 "},
  {4, " Ramdom Access R/W         "},
  {5, " Random Access Read only   "},
  {6, " Cyclic Access R/W         "},
  {7, " Cyclic Access Read only   "},
  {8, " Purse (Direct)            "},
  {9, " Purse (Cashback/decrement)"},
  {10, " Purse (Decrement)         "},
  {11, " Purse (Read Only)         "},
  {-1, " INVALID or UNKNOWN        "}
};

void
print_service_info(uint16 attr)
{
  unsigned int j;

  for (j = 0; j < sizeof(Sifo) / sizeof(*Sifo); j++) {
    if ((attr >> 1) == Sifo[j].c) {
      printf("%s", Sifo[j].d);
      break;
    }
  }

  if (!(attr & 1)) {
    printf(" (PROTECTED) ");
  }
  printf("\n");
}

void
dump_service(felica * f)
{
  uint8 b[16];
  int i, j, r;

  r = felica_search_service(f);
  if (r)
    return;

  /* print header */
  printf("# card IDm = ");
  mydump(f->IDm, 8);
  printf("\n# card PMm = ");
  mydump(f->PMm, 8);

  printf("\n# area num = %d", f->area_num);
  printf("\n# service num = %d", f->service_num);

  for (i = 0; i < f->area_num; i++) {
    printf("\n# AREA #%d = %04X (%05x)", i, f->area[i].code, f->area[i].attr);
  }
  printf("\n");

  /* dump services */
  for (i = 0; i < f->service_num; i++) {
    printf("# ");
    printf("%04X:%04x", f->service[i].code, f->service[i].attr);

    print_service_info(f->service[i].attr);

    if (f->service[i].attr & 1) {
      j = 0;
      while (!felica_read_single(f, f->service[i].bin, 0, j, b)) {
	printf("  ");
	printf("%04X:%04X:", f->service[i].code, j);
	mydump(b, 16);
	printf("\n");
	j++;
      }
    }
  }
}

void
enum_service(felica * f)
{
  felica *ff;
  uint16 resp[256];
  int i, n, r;

  n  = sizeof(resp)/sizeof(*resp);
  r = felica_request_system(f, &n, resp);
  if (r){
    printf("can't get service\n");
    return;
  }

  printf("  system num %d\n", n);

  for (i = 0; i < n; i++) {
    ff = felica_polling(f->p, resp[i], 0, 0);
    if (ff == NULL) {
      printf("can't get service  %d\n", i);
      break;
    }
    printf("# FELICA SYSTEM_CODE = %04X\n", resp[i]);
    dump_service(ff);
    free(ff);
  }
}

int
main(void)
{
  pasori *p;
  felica *f;

  p = pasori_open();
  if (!p) {
    printf("error\n");
    exit(-1);
  }
  pasori_init(p);
  f = felica_polling(p, FELICA_POLLING_ANY, 0, 0);
  if (f) {
    show_idminfo(f->IDm);
    enum_service(f);
  }
  free(f);
  pasori_close(p);

  return 0;
}


tests/pasori_test.c
/* $Id: pasori_test.c,v 1.5 2009-10-09 07:43:14 hito Exp $ */
#include <stdio.h>
#include <stdlib.h>

#include "config.h"
#include "libpafe.h"

int
test(pasori *p)
{
  unsigned char data[] = "test data.";
  int n = sizeof(data) - 1;
  
  printf("Echo test... ");
  if (pasori_test_echo(p, data, &n)) {
    printf("error!\n");
    return 1;
  } else {
    printf("success\n");
  }

  printf("EPROM test... ");
  if (pasori_test_eprom(p)) {
    printf("error!\n");
    return 1;
  } else {
    printf("success\n");
  }

  printf("RAM test... ");
  if (pasori_test_ram(p)) {
    printf("error!\n");
    return 1;
  } else {
    printf("success\n");
  }

  printf("CPU test... ");
  if (pasori_test_cpu(p)) {
    printf("error!\n");
    return 1;
  } else {
    printf("success\n");
  }

  printf("Polling test... ");
  if (pasori_test_polling(p)) {
    printf("error!\n");
    return 1;
  } else {
    printf("success\n");
  }
  return 0;
}

void
show_version(pasori *p)
{
  int v1, v2;

  if (pasori_version(p, &v1, &v2)) {
    printf("cannot get version.\n");
    return;
  }

  switch (pasori_type(p)) {
  case PASORI_TYPE_S310:
    printf("PaSoRi (RC-S310)\n firmware version %d.%02d\n", v1, v2);
    break;
  case PASORI_TYPE_S320:
    printf("PaSoRi (RC-S320)\n firmware version %d.%02d\n", v1, v2);
    break;
  case PASORI_TYPE_S330:
    printf("PaSoRi (RC-S330)\n firmware version %d.%02d\n", v1, v2);
    break;
  default:
    printf("unknown hardware.\n");
  }

}

int
main(void)
{
  pasori *p;
  int type;

  p = pasori_open();
  if (!p) {
    printf("error\n");
    exit(-1);
  }
  pasori_init(p);

  show_version(p);

  type = pasori_type(p);
  if ((type == PASORI_TYPE_S310 || type == PASORI_TYPE_S320) && test(p)) {
    pasori_close(p);
    return 1;
  }

  pasori_close(p);

  return 0;
}


src/pasori_command.h
/* $Id: pasori_command.h,v 1.3 2009-07-17 07:28:35 hito Exp $ */
#ifndef __PASORI_COMMAND_H
#define __PASORI_COMMAND_H

pasori *pasori_open(void);
int pasori_init(pasori *p);
void pasori_close(pasori *p);
int pasori_send(pasori *p, uint8 *data, int *size);
int pasori_recv(pasori *p, uint8 *data, int *size);
int pasori_packet_write(pasori *p, uint8 *data, int *size);
int pasori_packet_read(pasori * p, uint8 * data, int *size);
int pasori_write(pasori *p, uint8 *data, int *size);
int pasori_read(pasori *p, uint8 *data, int *size);
int pasori_reset(pasori * p);
int pasori_version(pasori *p, int *v1, int *v2);
int pasori_type(pasori *p);

int pasori_test(pasori *p, int code, uint8 *data, int *size, uint8 *rdata, int *rsize);
int pasori_test_echo(pasori *p, uint8 *data, int *size);
int pasori_test_eprom(pasori *p);
int pasori_test_ram(pasori *p);
int pasori_test_cpu(pasori *p);
int pasori_test_polling(pasori *p);
int pasori_list_passive_target(pasori *pp, unsigned char *payload, int *size);

void pasori_set_timeout(pasori *p, int timeout);

#ifdef DEBUG
#define Log printf
#else
#define Log 1 ? (void) 0 : printf
#endif

#define L8(x) (x & 0xff)
#define H8(x) ((x >> 8) & 0xff)

#endif


src/libpafe.h
/* $Id: libpafe.h,v 1.8 2009-10-31 08:54:53 hito Exp $ */
#ifndef __LIBPASORI_H
#define __LIBPASORI_H

/* DEFINE types */

#ifndef uint8
typedef unsigned char uint8;
#endif

#ifndef uint16
typedef unsigned short int uint16;
#endif

/* DEFINE pasori */

enum PASORI_TYPE {
  PASORI_TYPE_S310,
  PASORI_TYPE_S320,
  PASORI_TYPE_S330,
};

struct tag_pasori;

typedef struct tag_pasori pasori;
/*~DEFINE pasori */

typedef struct _ferica_area
{
  uint16 code, attr, bin;
  struct _ferica_area *next;
} felica_area;

/* DEFINE felica */

struct tag_felica
{
  pasori *p;
  uint16 systemcode;
  uint8 IDm[8];
  uint8 PMm[8];
  uint16 area_num;
  felica_area area[256];
  uint16 service_num;
  felica_area service[256];
  struct tag_felica *next;
};
typedef struct tag_felica felica;

struct tag_felica_block
{
  uint8 data[8];
};
typedef struct tag_felica_block felica_block;

typedef struct _felica_block_info
{
  uint16 service;
  uint8 mode;
  uint16 block;
} felica_block_info;

/*~DEFINE felica */

#include "pasori_command.h"
#include "felica_command.h"


/* DEFINE constants */

#define FELICA_IDM_LENGTH 8
#define FELICA_PMM_LENGTH 8
#define FELICA_BLOCK_LENGTH 16

#define FELICA_AREA_NUM_MAX 121

#define FELICA_POLLING_ANY   0xffff
#define FELICA_POLLING_SUICA 0x0003
#define FELICA_POLLING_EDY   0xfe00


#define FELICA_SERVICE_SUICA_IN_OUT  0x108f
#define FELICA_SERVICE_SUICA_HISTORY 0x090f
#define FELICA_SERVICE_IRUCA_HISTORY 0xde80
#define FELICA_SERVICE_EDY_NUMBER  0x110b
#define FELICA_SERVICE_EDY_HISTORY 0x170f


#define FELICA_CMD_POLLING 0
#define FELICA_ANS_POLLING 1

#define FELICA_CMD_REQUEST_SERVICE 2
#define FELICA_ANS_REQUEST_SERVCCE 3

#define FELICA_CMD_REQUEST_RESPONSE 4
#define FELICA_ANS_REQUEST_RESPONSE 5

#define FELICA_CMD_READ_WITHOUT_ENCRYPTION 6
#define FELICA_ANS_READ_WITHOUT_ENCRYPTION 7

#define FELICA_CMD_WRITE_WITHOUT_ENCRYPTION 8
#define FELICA_ANS_WRITE_WITHOUT_ENCRYPTION 9

#define FELICA_CMD_AUTHENTICATION1 0x10
#define FELICA_ANS_AUTHENTICATION1 0x11

#define FELICA_CMD_AUTHENTICATION2 0x12
#define FELICA_ANS_AUTHENTICATION2 0x13

#define FELICA_CMD_READ_FROM_SECURE_FILE 0x14
#define FELICA_ANS_READ_FROM_SECURE_FILE 0x15

#define FELICA_CMD_WRITE_TO_SECURE_FILE 0x16
#define FELICA_ANS_WRITE_TO_SECURE_FILE 0x17


#define FELICA_CMD_SEARCH_SERVICE_CODE 0x0a
#define FELICA_ANS_SEARCH_SERVICE_CODE 0x0b

#define FELICA_CMD_REQUEST_SYSTEM 0x0c
#define FELICA_ANS_REQUEST_SYSTEM 0x0d


#define PASORI2_CMD_SELF_DIAGNOSIS 0x52
#define PASORI2_ANS_SELF_DIAGNOSIS 0x53

#define PASORI2_CMD_GET_FIRMWARE_VERSION 0x58
#define PASORI2_ANS_GET_FIRMWARE_VERSION 0x59

#define PASORI2_CMD_SEND_PACKET 0x5c
#define PASORI2_ANS_SEND_PACKET 0x5d

#define PASORI2_DIAG_COMMUNICATION_LINE_TEST 0x00
#define PASORI2_DIAG_EEPROM_TEST 0x01
#define PASORI2_DIAG_RAM_TEST 0x02
#define PASORI2_DIAG_CPU_FUNCTION_TEST 0x03
#define PASORI2_DIAG_CPU_FANCTION_TEST 0x03
#define PASORI2_DIAG_POLLING_TEST_TO_CARD 0x10


#define PASORI_ERR_PARM 1
#define PASORI_ERR_MEM  2
#define PASORI_ERR_COM  3
#define PASORI_ERR_DATA 4
#define PASORI_ERR_CHKSUM 5
#define PASORI_ERR_FORMAT 6
#define PASORI_ERR_TYPE 7

/*~DEFINE constants */

#endif
